src/java/org/apache/cassandra/cql3/PasswordObfuscator.java

```diff
@@ -62,10 +62,15 @@ public static String obfuscate(String query, RoleOptions opts)
             return query;
 
         Optional<String> pass = opts.getPassword();
+        if (!pass.isPresent() || pass.get().isEmpty())
+            pass = opts.getHashedPassword();
         if (!pass.isPresent() || pass.get().isEmpty())
             return query;
 
-        // match new line, case insensitive (?si), and PASSWORD_TOKEN up to the actual password greedy. Group that and replace the password
-        return query.replaceAll("((?si)"+ PASSWORD_TOKEN + ".+?)" + pass.get(), "$1" + PasswordObfuscator.OBFUSCATION_TOKEN);
+        // Regular expression:
+        //  - Match new line and case insensitive (?si), and PASSWORD_TOKEN with greedy mode up to the start of the actual password and group it.
+        //  - Quote the password between \Q and \E so any potential special characters are ignored
+        //  - Replace the match with the grouped data + the obfuscated token
+        return query.replaceAll("((?si)"+ PASSWORD_TOKEN + ".+?)\\Q" + pass.get() + "\\E", "$1" + PasswordObfuscator.OBFUSCATION_TOKEN);
```

Comment:
Nit: I would split the arguments for better readability, feel free to ignore if you don't agree:
```suggestion
        return query.replaceAll("((?si)"+ PASSWORD_TOKEN + ".+?)\\Q" + pass.get() + "\\E",
                                "$1" + PasswordObfuscator.OBFUSCATION_TOKEN);
```